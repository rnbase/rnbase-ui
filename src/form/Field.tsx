import React, { Children } from 'react';
import { Image, StyleProp, StyleSheet, Text, View, ViewStyle } from 'react-native';

import { Theme, Themed, withTheme } from '../theming';

interface FieldProps {
  children: React.ReactElement;
  error?: React.ReactNode;
  touch?: boolean;
  style?: StyleProp<ViewStyle>;
  label?: string;
}

const Field: React.FC<Themed<typeof createStyleSheet, FieldProps>> = ({
  theme: { styles },
  error,
  touch = false,
  label,
  children,
}) => (
  <View style={styles.root}>
    {label && (
      <View style={styles.heading}>
        <Text style={styles.label} numberOfLines={1}>
          {label}
        </Text>
        {error && touch && (
          <Text style={[styles.label, styles.labelError]} numberOfLines={1}>
            {error}
          </Text>
        )}
      </View>
    )}
    <View style={styles.field}>
      {React.cloneElement(Children.only(children), {
        style: children.props.style
          ? StyleSheet.flatten([styles.input, children.props.style])
          : styles.input,
      })}
      {error && touch && <Image style={styles.errorIcon} source={require('../assets/error.png')} />}
    </View>
  </View>
);

const createStyleSheet = ({ colors, fonts }: Theme) =>
  StyleSheet.create({
    root: {
      marginBottom: 20,
    },
    heading: {
      alignItems: 'center',
      flexDirection: 'row',
      justifyContent: 'space-between',
    },
    label: {
      ...fonts.semibold,
      height: 11,
      fontSize: 11,
      lineHeight: 11,
      flexShrink: 1,
      marginBottom: 8,
      color: colors.black,
      textTransform: 'uppercase',
      backgroundColor: colors.white,
    },
    labelError: {
      flexShrink: 0,
      marginLeft: 10,
      color: colors.red,
      textTransform: 'none',
    },
    field: {
      borderRadius: 5,
      alignItems: 'center',
      flexDirection: 'row',
      backgroundColor: colors.gray6,
    },
    input: {
      ...fonts.normal,
      height: 50,
      flexGrow: 1,
      fontSize: 18,
      flexShrink: 1,
      borderWidth: 0,
      paddingVertical: 0,
      paddingHorizontal: 15,
      backgroundColor: colors.transparent,
    },
    errorIcon: {
      width: 21,
      height: 21,
      marginRight: 8,
    },
  });

export default withTheme(Field, createStyleSheet, 'Field');
