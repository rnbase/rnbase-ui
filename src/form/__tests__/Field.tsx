import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

/**
 * Under test
 */
import Field from '../Field';

type Props = React.ComponentProps<typeof Field>;

const Input = (props: any) => React.createElement('Input', props);

const createElement = (props?: Partial<Props>, style?: any) => (
  <Field {...props}>
    <Input style={style} />
  </Field>
);

const createRenderer = (props?: Partial<Props>, style?: any) =>
  TestRenderer.create(createElement(props, style));

it('should render normally', () => {
  expect(createRenderer()).toMatchSnapshot();
});

it('should render with label', () => {
  expect(createRenderer({ label: 'label' })).toMatchSnapshot();
});

it('should render with error', () => {
  expect(createRenderer({ label: 'label', touch: true, error: 'error' })).toMatchSnapshot();
});

it('should render with custom input styles', () => {
  expect(createRenderer({}, { color: 'red' })).toMatchSnapshot();
});
