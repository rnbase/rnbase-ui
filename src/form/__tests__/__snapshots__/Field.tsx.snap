// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render normally 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 20,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f7",
          "borderRadius": 5,
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <Input
      style={
        Object {
          "backgroundColor": undefined,
          "borderWidth": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "System",
          "fontSize": 18,
          "fontWeight": "400",
          "height": 50,
          "includeFontPadding": false,
          "paddingHorizontal": 15,
          "paddingVertical": 0,
          "textAlignVertical": "center",
        }
      }
    />
  </View>
</View>
`;

exports[`should render with custom styles 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 20,
      },
      Object {
        "backgroundColor": "red",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        Object {
          "backgroundColor": "green",
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "color": "#000000",
            "flexShrink": 1,
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "height": 11,
            "includeFontPadding": false,
            "lineHeight": 11,
            "marginBottom": 8,
            "textAlignVertical": "center",
            "textTransform": "uppercase",
          },
          Object {
            "color": "blue",
          },
        ]
      }
    >
      label
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "color": "#000000",
            "flexShrink": 1,
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "height": 11,
            "includeFontPadding": false,
            "lineHeight": 11,
            "marginBottom": 8,
            "textAlignVertical": "center",
            "textTransform": "uppercase",
          },
          Object {
            "color": "#ff3b30",
            "flexShrink": 0,
            "marginLeft": 10,
            "textTransform": "none",
          },
          Object {
            "color": "magenta",
          },
        ]
      }
    >
      error
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f7",
          "borderRadius": 5,
          "flexDirection": "row",
        },
        Object {
          "backgroundColor": "orange",
        },
      ]
    }
  >
    <Input
      style={
        Object {
          "backgroundColor": undefined,
          "borderWidth": 0,
          "color": "yellow",
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "System",
          "fontSize": 18,
          "fontWeight": "400",
          "height": 50,
          "includeFontPadding": false,
          "paddingHorizontal": 15,
          "paddingVertical": 0,
          "textAlignVertical": "center",
        }
      }
    />
    <Image
      source={
        Object {
          "uri": "error.png",
        }
      }
      style={
        Array [
          Object {
            "height": 21,
            "marginRight": 8,
            "width": 21,
          },
          Object {
            "tintColor": "cyan",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`should render with error 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 20,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "color": "#000000",
            "flexShrink": 1,
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "height": 11,
            "includeFontPadding": false,
            "lineHeight": 11,
            "marginBottom": 8,
            "textAlignVertical": "center",
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      label
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "color": "#000000",
            "flexShrink": 1,
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "height": 11,
            "includeFontPadding": false,
            "lineHeight": 11,
            "marginBottom": 8,
            "textAlignVertical": "center",
            "textTransform": "uppercase",
          },
          Object {
            "color": "#ff3b30",
            "flexShrink": 0,
            "marginLeft": 10,
            "textTransform": "none",
          },
          undefined,
        ]
      }
    >
      error
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f7",
          "borderRadius": 5,
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <Input
      style={
        Object {
          "backgroundColor": undefined,
          "borderWidth": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "System",
          "fontSize": 18,
          "fontWeight": "400",
          "height": 50,
          "includeFontPadding": false,
          "paddingHorizontal": 15,
          "paddingVertical": 0,
          "textAlignVertical": "center",
        }
      }
    />
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/error.png",
        }
      }
      style={
        Array [
          Object {
            "height": 21,
            "marginRight": 8,
            "width": 21,
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`should render with label 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 20,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "color": "#000000",
            "flexShrink": 1,
            "fontFamily": "System",
            "fontSize": 11,
            "fontWeight": "600",
            "height": 11,
            "includeFontPadding": false,
            "lineHeight": 11,
            "marginBottom": 8,
            "textAlignVertical": "center",
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      label
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f7",
          "borderRadius": 5,
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <Input
      style={
        Object {
          "backgroundColor": undefined,
          "borderWidth": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "System",
          "fontSize": 18,
          "fontWeight": "400",
          "height": 50,
          "includeFontPadding": false,
          "paddingHorizontal": 15,
          "paddingVertical": 0,
          "textAlignVertical": "center",
        }
      }
    />
  </View>
</View>
`;
